[AIRFLOW-2203] Dag import speed

wongwill86 opened the pull request:
### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
   - https://issues.apache.org/jira/browse/AIRFLOW-2203


### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

Importing DAGS are terribly slow and operates at polynomial time (or worse). These changes bring huge performance boost for users who need have deep and wide dags. In the new toy example of a very basic Fan-In/Fan-Out x4 DAG found in `tests/dags/test_performance.py`, fails to compile in any reasonable timeframe using the original code. The following optimizations are employed:

1. Defer detecting cycles of DAGs. Currently `detect_downstream_cycle` traverses the entire graph every time a task is added. Updated to call cycle detection only once when the DAG is collected by the `DagBag`. This is crucial because it reduces runtime from polynomial to near linear. Commit `8e8e735`
2. Remove useless calls to `dag.tasks` in add_task command which iterates through every task needlessly for *every* task added. ~~Commit `013a7d2`~~
3. Cache expensive `inspect.signature` call. Currently being called at every `BaseOperator/Operator` construction. We should be able to cache this because the function signature of the constructor isn't really expected to change. ~~Commit `a7b230e`~~
4. ~~Use default `Resources` object for `BaseOperator` constructor. Currently every time an operator is constructed without specifying the `resources` parameter, a brand new `Resources` object is created that calls `configuration.getint()` 4x (to create `cpu/ram/disk/gpuResource`). A default declared `Resources` object can be shared for most tasks that don't need this functionality. Commit `6da941e`~~ Set default values for `Resources` object to be pulled from configuration. This changes from needing 4x configuration loads for every operator created to only 4x configuration loads to set the default values once.
   * ~~*WARNING* potential breaking change! Because the default Resource object is shared among all operators (when none is specified). `Resources`'s internal `Resource` objects are assumed to be immutable (hence the change to getter access only). If a task needs to have its `resources` modified, the user must create a new one manually.~~
5. Stop trigger/weight rules from recalculating the classes' static variables. ~~Commit `5d68744`~~
6. Maintain tasks' `_upstream_task_ids` and `_downstream_task_ids` as sets instead of lists. Currently everytime a task is added, it iterates through the entire list to see if the task has already been added. There shouldn't be a need to do this.
    * *WARNING* potential breaking change. The order in which a task is `set_downstream`/`set_upstream` is no longer preserved when retrieving the task list anymore. Order shouldn't have been an expected behavior to begin with, but just in case some people relied on that for some reason, they should be made aware.

An idea of the performance on my very old Haswell Laptop (in seconds):
1. Deferred Cycle Detection- Compared with no cycle detection. Only adds a little more time and at least the DAG can now be loaded (compared to original)


Width | Tasks | Original | No Cycle Detection | Deferred Cycle Detection | All changes
-- | -- | -- | -- | -- | --
25 | 105 | 6.67 | 0.312 | 0.373 | 0.234
50 | 205 | 199 | 0.352 | 0.436 | 0.329
100 | 405 | > 300 | 0.416 | 0.485 | 0.323
200 | 805 | >300 | 0.65 | 0.648 | 0.367
400 | 1605 | > 300 | 0.87 | 0.996 | 0.45
800 | 3205 | > 300 | 1.53 | 1.77 | 0.534
1600 | 6405 | > 300 | 3.24 | 3.86 | 0.773
3200 | 12805 | > 300 | 9.47 | 11.25 | 1.12
6400 | 25605 | > 300 | 36 | 42.43 | 2.18
12800 | 51205 | > 300 | 151 | 163 | 3.77
51200 | 204805 | > 300 |  >151 | > 163 | 14.75
102300 | 409605 | > 300| > 151  | > 163 | 28.8

Full benchmarks as each feature added can be found here:
https://docs.google.com/spreadsheets/d/1_39JRdPpXlWbWVM9jRVn7kBaIK7TshJllr5COC30Xeo/edit?usp=sharing


### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

- tests/utils/test_trigger_rule.py:TestTriggerRule
- tests/utils/test_weight_rule.py:TestWeightRule
- tests/utils/test_decorators.py:ApplyDefaultTest
- test/utils.py:OperatorResourcesTest.test_unable_to_change_resource
- tests/dags/test_performance.py
- tests/models.py:DagTest.test_cycle
- tests/models.py:DagBagTest.test_load_subdags
- tests/models.py:DagBagTest.test_skip_cycle_dags
- tests/models.py:DagBagTest.test_load_large_dag

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
   1. Subject is separated from body by a blank line
   2. Subject is limited to 50 characters
   3. Subject does not end with a period
   4. Subject uses the imperative mood ("add", not "adding")
   5. Body wraps at 72 characters
   6. Body explains "what" and "why", not "how"

~~Not yet, but I will. There are 6 different improvements that can be applied separately. I wasn't sure if they should each be individual PR's but they each contribute significantly to the performance boost. I kept them separate in case I needed to pull an individual change out of this PR. If everything checks out, I will squash them into one single commit.~~

As discussed on the PR below, I kept each of these features separate in a different commit (all prepended with [AIRFLOW-2203])

- [X] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

saguziel commented on airflow/models.py:
```diff
@@ -2207,7 +2237,7 @@ def __init__(
             on_retry_callback=None,
             trigger_rule=TriggerRule.ALL_SUCCESS,
-            resources=None,
+            resources=Resources(),
```
the Resources object will be shared across invocations here

wongwill86 replied:
Yes that's a potential breaking change. I couldn't find where a task's `resources` was actually being modified anywhere after construction and I didn't quite understand its usage for when that would actually happen. I only saw this being accessed in `cgroup_task_runner`. Did I miss a use case here? Additionally, I added a change to prevent modifying an existing Resource object as well.

ashb replied:
I _think_ that this change is okay, especially with you having made the object's properties read-only.  It might be worth a comment somewhere saying "shared by design" with a small bit of rationale

saguziel replied:
I’m gonna vote for removal. Even if read only now, it may change in the future. Also since it’s usually a bad pattern, I would call it bad style/confusing.

wongwill86 replied:
Just to make sure we're on the same page here this was a pretty big performance sink with the 4x configuration.getInt requests for every new operator.

Width | Tasks | New Resource Object | Static Resources Object
-- | -- | -- | --
12800 | 51205 | 4.85 | 2.7
51200 | 204805 | 18 | 10.3

I'm ok with pulling this commit out:
[Share default Resources object when no Resources specified 6da941e](https://github.com/apache/incubator-airflow/pull/3116/commits/6da941e808a2e1e858e1c20ecf0df70b3bd3d932)

That said, users aren't completely forced into this performance sink, just a bit clunky to have ask users to manually share a default Resources object if they care about performance of very large dags. I can add a comment to advise this if they are.

As an alternative, what do you think about this diff in `operators_resources.py`? at least then we don't have to load the configuration 4x with every new operator (interpolating/checking env vars, checking `airflow.cfg`, etc ...)
```
+++ b/airflow/utils/operator_resources.py
@@ -88,7 +88,9 @@ class GpuResource(Resource):
 class Resources(object):
-    def __init__(self, cpus=None, ram=None, disk=None, gpus=None):
-        if cpus is None:
-            cpus = configuration.getint('operators', 'default_cpus')
-        if ram is None:
-            ram = configuration.getint('operators', 'default_ram')
-        if disk is None:
-            disk = configuration.getint('operators', 'default_disk')
-        if gpus is None:
-            gpus = configuration.getint('operators', 'default_gpus')
+    def __init__(self,
+                 cpus=configuration.getint('operators', 'default_cpus'),
+                 ram=configuration.getint('operators', 'default_ram'),
+                 disk=configuration.getint('operators', 'default_disk'),
+                 gpus=configuration.getint('operators', 'default_gpus')
+                 ):

```

Thanks for reviewing! 
Let me know what you would like me to do! 😃

ashb replied:
Yeah, I think this change sounds like a good compromise.

(This now makes me wonder just how slow ConfigParser is and if it's causing any other bottlenecks in Airflow.... 😟 )

saguziel commented on airflow/models.py:
```diff
@@ -335,10 +339,17 @@ def process_file(self, filepath, only_if_updated=True, safe_mode=True):
+                        self.bag_dag(dag, parent_dag=dag, root_dag=dag)
+                        found_dags.append(dag)
+                        found_dags += dag.subdags
+                    except Exception as e:
```
Doesn't this happen somewhere else before?

wongwill86 replied:
If you're referring to the exception handling, yes this is the same exception handling as a few lines above:
https://github.com/apache/incubator-airflow/pull/3116/files#diff-a32a363fa616685db3bfefba947535b2R300

Prior to this PR, the cycle detection exception was raised immediately when a task was added to a DAG and an error would be thrown directly when the dag was imported.

Now with this PR, the cycle detection exception is raised when bag_dag is called (i.e. before a dag can be added to the dagbag).

Is that what you're referring to?

Fokko replied:
The exception is really broad, can we make this more specific?

wongwill86 replied:
Yes, added new `AirflowDagCycleException`

saguziel commented on airflow/models.py:
```diff
@@ -381,20 +392,39 @@ def kill_zombies(self, session=None):
+
+            self.dags[dag.dag_id] = dag
+            self.log.debug('Loaded DAG {dag}'.format(**locals()))
+        except Exception as e:
```
Is this a change in behavior also?

wongwill86 replied:
yes the new behavior is that the bag_dag method may throw an exception if we try to bag a dag with cycles in it. Otherwise, the logic should remain the same. 

To do this, additional logic is introduced to ensure that the state of the dagbag is reverted (i.e. undo `self.dags[dag.dag_id] = dag`). This is handled by backing out of the subdag DFS and at the root dag, remove all known subdags ( `dag.subdag` actually returns all DFS subdags already). The root dag is now only added to the dagbag when all subdags are successfully added so we don't have to remove it from the dagbag manually.

Fokko replied:
The exception is really broad, can we make this more specific?

Fokko commented on airflow/models.py:
```diff
@@ -2281,13 +2311,17 @@ def __init__(
-        self._upstream_task_ids = []
-        self._downstream_task_ids = []
+        self._upstream_task_ids = set()
+        self._downstream_task_ids = set()
```
I like this change

Fokko commented on airflow/models.py:
```diff
@@ -2771,13 +2790,13 @@ def __repr__(self):
-    def append_only_new(self, l, item):
-        if any([item is t for t in l]):
+    def add_only_new(self, s, item):
+        if item in s:
```
Now we are changing this lines, can we make the variable names more meaningful? I dislike the use of `t`, `l` and `s`. Cheers

wongwill86 replied:
really bothered me too! updated!

ashb commented on airflow/utils/decorators.py:
```diff
@@ -30,6 +30,9 @@
 
 def apply_defaults(func):
+
+    sig_cache = dict()
```
Is this in the right place to cache more than a single function call?

```
@apply_defaults
def a():
    pass

@apply_defaults
def b():
    pass

```

I _think_ that `a` and `b` get a different sig_cache. Is that okay? expected? desired? behaviour

jgao54 replied:
I _think_ this is okay, since @apply_default is only applied to `__init__`.

It is also consistent with prior the change (i.e. a and b get different sig_cache)

wongwill86 replied:
yea i think this should be the expected and desired result.
expected and desired b/c like @jgao54 mentioned, each __init__ *should* have been set up with its own wrapper as was the previous behavior.

It is required to be here but usage of the dict is slightly disingenuous. It's not here to actually cache more than a single function but to solve the closure problem of the `wrapper` function which refers to the out of scope sig_cache. If we did something like `sig_cache = (sig, non_optional_args)` this would merely rebind the sig_cache inside `wrapper`'s scope and not actually work at all.

if the dict is too confusing, should I change it to a private class with a private class instead?
something along the lines of:
```
def apply_defaults(func):
    class SignatureCache(object):
        def __init__(self):
            self.value = None
     
    sig_cache = SignatureCache()
... 
    def wrapper(...
...
         if sig_cache.value :
+            (sig, non_optional_args) = sig_cache.value 
+        else:
+            sig = signature(func)
+            non_optional_args = {
+                name for (name, param) in sig.parameters.items()
+                if param.default == param.empty and
+                param.name != 'self' and
+                param.kind not in (param.VAR_POSITIONAL, param.VAR_KEYWORD)}
+            sig_cache.value= (sig, non_optional_args)
```

Would that be preferred?

jgao54 replied:
+1 on the suggestion, and may be worth commenting inline as you have explained above :)

ashb commented on tests/models.py:
```diff
@@ -857,6 +986,323 @@ def test_get_dag_fileloc(self):
+
+        # Validate correctness
+        # None of the dags should be found
+        self.validate_dags(testDag, found_dags, dagbag, should_be_found=False)
```
Is it worth checking that we record the import error in this DAG so that a warning show up in the webserver? (I don't actually know the mechanism by which that happens mind...)

wongwill86 replied:
yup this should show up in the webserver. At some point, `DagBag.import_errors` gets transformed and read from the view layer on the webpage.

ashb commented on tests/models.py:
```diff
@@ -865,6 +1311,14 @@ def test_process_file_with_none(self):
 
         self.assertEqual([], dagbag.process_file(None))
 
+    @timed(2)
```
I am very wary of this -- given we run our tests on Travis this seems like a recipe for transient failures.

wongwill86 replied:
i agree. i'll remove it. FWIW there're are a bunch of tests that also have brittle performance behavior:
see https://github.com/wongwill86/incubator-airflow/blob/master/tests/jobs.py#L2524 (test_scheduler_run_duration) and a few other tests in that file. the timeout in this test covers a bit more than simply bagging a dag too.

Are there any other ways people know of how to track performance statistics from build to build, like coveralls for performance? Perhaps not for this PR but at some other point.

ashb commented on tests/utils.py:
```diff
@@ -20,10 +20,12 @@
 import unittest
 
-import airflow.utils.logging
+# import airflow.utils.logging
```
Remove rather than comment out please

wongwill86 replied:
totally unintentional here. added this back in.

feng-tao commented on airflow/models.py:
```diff
@@ -63,7 +63,9 @@
 from airflow import configuration
-from airflow.exceptions import AirflowException, AirflowSkipException, AirflowTaskTimeout
+from airflow.exceptions import (
+    AirflowException, AirflowSkipException, AirflowTaskTimeout, AirflowDagCycleException
```
alphabetical import order

wongwill86 committed:
[AIRFLOW-2203] Defer cycle detection

Moved from adding_task to when dag is being bagged.
This changes import dag runtime from polynomial to somewhat linear.

wongwill86 committed:
[AIRFLOW-2203] Remove Useless Commands.

self.tasks is a temp list gen from self.task_dict. no reason to append
to it

wongwill86 committed:
[AIRFLOW-2203] Cache signature in apply_defaults

Cache inspect.signature for the wrapper closure to avoid calling it at
every decorated invocation. This is separate sig_cache created per
decoration, i.e. each function decorated using apply_defaults will have
a different sig_cache.

wongwill86 committed:
[AIRFLOW-2203] Speed up Operator Resources

Set default values of Resources loaded from the configuration to prevent
4x Config lookups for every task created.

wongwill86 committed:
[AIRFLOW-2203] Cache static rules (trigger/weight)

No need to recalculate them everytime just to see if they are valid

wongwill86 committed:
[AIRFLOW-2203] Store task ids as sets not lists

Massively improve performance by using sets to represent a task's
upstream and downstream task ids.
